File: web_helper/frontend/src/shared/model/hierarchy.ts
/**
 * Hierarchical Node Graph Types for Simulink-style visualization.
 *
 * Matches backend models/hierarchy.py schema.
 */

// =============================================================================
// Enums
// =============================================================================

export type HierarchyLevel = "method" | "component" | "layer" | "operation";

export type ComponentCategory =
  | "model"
  | "optimizer"
  | "scheduler"
  | "loss"
  | "dataset"
  | "dataloader"
  | "transform"
  | "metric"
  | "sampler"
  | "checkpoint"
  | "callback"
  | "logger"
  | "agent"
  | "global"
  | "unknown";

/**
 * Method categories for Agent methods.
 * Matches backend hierarchy.py MethodCategory enum.
 */
export type MethodCategory =
  | "setup"       // setup() - initialization
  | "train"       // train_step() - training logic
  | "validation"  // val_step() - validation logic
  | "lifecycle"   // on_train_start, on_train_end, etc.
  | "utility";    // Other helper methods

export type FlowType =
  | "tensor"
  | "parameters"
  | "gradient"
  | "control"
  | "config"
  | "reference";

export type OriginType =
  | "create_call"
  | "assignment"
  | "method_call"
  | "unpack"
  | "input"
  | "output";

export type ImplSource = "yaml" | "default" | "hardcode";

/**
 * Value source for properties (5-state system).
 * Matches backend hierarchy.py ValueSource enum.
 */
export const ValueSource = {
  REQUIRED: "required",   // No default value, must be provided
  CONFIG: "config",       // Value from YAML config file
  DEFAULT: "default",     // Has default value in code
  HARDCODE: "hardcode",   // Fixed in code, not configurable
  CONNECTED: "connected", // Value from connected node edge
} as const;

export type ValueSource = typeof ValueSource[keyof typeof ValueSource];

/**
 * Property information with value source for UI display.
 * Used in SmartNodeView to show properties with clear source indicators.
 */
export interface PropertyInfo {
  name: string;
  value: unknown;
  source: ValueSource;
  is_grid_search?: boolean;
  grid_count?: number;
  default_value?: unknown;
  description?: string;
}

/**
 * Summary counts of properties by source for badge display.
 * Shows "⚠️ 2 required, 3 config" on node surface.
 */
export interface PropertySummary {
  required_count: number;
  config_count: number;
  default_count: number;
  hardcode_count: number;
  connected_count: number;
}

// =============================================================================
// Types
// =============================================================================

export interface SourceLocation {
  file: string;
  line: number;
  column?: number;
  end_line?: number;
  end_column?: number;
}

export interface NodeOrigin {
  type: OriginType;
  create_path?: string[];
  code_snippet: string;
  source?: SourceLocation;
  impl_source?: ImplSource;
}

export interface DrillTarget {
  type: "component" | "layer";
  path: string;
}

export interface Port {
  name: string;
  type: string;
  kind?: "exec" | "data";
}

export interface HierarchyPath {
  level: HierarchyLevel;
  label: string;
  node_id: string;
  graph_id: string;
  // Optional: Category for visual styling (e.g., "model", "optimizer")
  category?: ComponentCategory;
  // Optional: Implementation name (e.g., "resnet18")
  implementation?: string;
}

export interface HierarchyNode {
  id: string;
  label: string;
  level: HierarchyLevel;

  // Code origin
  origin?: NodeOrigin;

  // Drill-down support
  can_drill: boolean;
  drill_target?: DrillTarget;

  // Visual category
  category?: ComponentCategory;

  // Ports
  inputs: Port[];
  outputs: Port[];

  // Source location
  source?: SourceLocation;

  // Properties with ValueSource (Phase 1)
  properties?: PropertyInfo[];
  property_summary?: PropertySummary;

  // Additional metadata
  metadata?: Record<string, unknown>;
}

export interface CodeFlowEdge {
  id: string;
  source_node: string;
  source_port: string;
  target_node: string;
  target_port: string;

  // Flow type
  flow_type: FlowType;
  edge_type?: "execution" | "data";

  // Variable name
  variable_name?: string;

  // Extraction source
  extracted_from?: "setup" | "train_step";

  // Creation order (for setup phase: 1, 2, 3, ...)
  sequence_index?: number;

  // Source location
  source?: SourceLocation;

  // Additional metadata
  metadata?: Record<string, unknown>;
}

export interface Hierarchy {
  parent_graph_id?: string;
  parent_node_id?: string;
  depth: number;
  path: HierarchyPath[];
}

export interface HierarchicalNodeGraph {
  id: string;
  label: string;
  level: HierarchyLevel;

  // Hierarchy context
  hierarchy: Hierarchy;

  // Graph content
  nodes: HierarchyNode[];
  edges: CodeFlowEdge[];

  // Agent name
  agent_name?: string;

  // Source file
  source_file?: string;

  // Code coverage info (for 1:1 code-node mapping)
  method_range?: SourceLocation;
  covered_lines?: number[];
  uncovered_lines?: number[];

  // Additional metadata
  metadata?: Record<string, unknown>;
}

// =============================================================================
// API Response Types
// =============================================================================

export interface HierarchyResponse {
  success: boolean;
  data?: HierarchicalNodeGraph;
  error?: string;
}

// =============================================================================
// Navigation State
// =============================================================================

export interface NavigationState {
  agentName: string;
  currentPath: string;
  graph: HierarchicalNodeGraph | null;
  history: Array<{
    path: string;
    graph: HierarchicalNodeGraph;
    viewport?: { x: number; y: number; zoom: number };
  }>;
}

// =============================================================================
// Theme Types
// =============================================================================

export interface CategoryTheme {
  // Legacy aliases used across the Builder UI
  bg: string;
  badge: string;
  background: string;
  border: string;
  text: string;
  icon?: string;
}

export const CATEGORY_THEMES: Record<ComponentCategory, CategoryTheme> = {
  model: {
    bg: "bg-blue-50 dark:bg-blue-950/30",
    badge: "bg-blue-50 dark:bg-blue-950/30",
    background: "bg-blue-50 dark:bg-blue-950/30",
    border: "border-blue-200 dark:border-blue-800",
    text: "text-blue-700 dark:text-blue-300",
    icon: "cube",
  },
  optimizer: {
    bg: "bg-green-50 dark:bg-green-950/30",
    badge: "bg-green-50 dark:bg-green-950/30",
    background: "bg-green-50 dark:bg-green-950/30",
    border: "border-green-200 dark:border-green-800",
    text: "text-green-700 dark:text-green-300",
    icon: "trending-up",
  },
  loss: {
    bg: "bg-red-50 dark:bg-red-950/30",
    badge: "bg-red-50 dark:bg-red-950/30",
    background: "bg-red-50 dark:bg-red-950/30",
    border: "border-red-200 dark:border-red-800",
    text: "text-red-700 dark:text-red-300",
    icon: "target",
  },
  dataset: {
    bg: "bg-purple-50 dark:bg-purple-950/30",
    badge: "bg-purple-50 dark:bg-purple-950/30",
    background: "bg-purple-50 dark:bg-purple-950/30",
    border: "border-purple-200 dark:border-purple-800",
    text: "text-purple-700 dark:text-purple-300",
    icon: "database",
  },
  dataloader: {
    bg: "bg-indigo-50 dark:bg-indigo-950/30",
    badge: "bg-indigo-50 dark:bg-indigo-950/30",
    background: "bg-indigo-50 dark:bg-indigo-950/30",
    border: "border-indigo-200 dark:border-indigo-800",
    text: "text-indigo-700 dark:text-indigo-300",
    icon: "loader",
  },
  transform: {
    bg: "bg-amber-50 dark:bg-amber-950/30",
    badge: "bg-amber-50 dark:bg-amber-950/30",
    background: "bg-amber-50 dark:bg-amber-950/30",
    border: "border-amber-200 dark:border-amber-800",
    text: "text-amber-700 dark:text-amber-300",
    icon: "wand",
  },
  metric: {
    bg: "bg-cyan-50 dark:bg-cyan-950/30",
    badge: "bg-cyan-50 dark:bg-cyan-950/30",
    background: "bg-cyan-50 dark:bg-cyan-950/30",
    border: "border-cyan-200 dark:border-cyan-800",
    text: "text-cyan-700 dark:text-cyan-300",
    icon: "bar-chart",
  },
  scheduler: {
    bg: "bg-orange-50 dark:bg-orange-950/30",
    badge: "bg-orange-50 dark:bg-orange-950/30",
    background: "bg-orange-50 dark:bg-orange-950/30",
    border: "border-orange-200 dark:border-orange-800",
    text: "text-orange-700 dark:text-orange-300",
    icon: "clock",
  },
  sampler: {
    bg: "bg-pink-50 dark:bg-pink-950/30",
    badge: "bg-pink-50 dark:bg-pink-950/30",
    background: "bg-pink-50 dark:bg-pink-950/30",
    border: "border-pink-200 dark:border-pink-800",
    text: "text-pink-700 dark:text-pink-300",
    icon: "shuffle",
  },
  checkpoint: {
    bg: "bg-slate-50 dark:bg-slate-950/30",
    badge: "bg-slate-50 dark:bg-slate-950/30",
    background: "bg-slate-50 dark:bg-slate-950/30",
    border: "border-slate-200 dark:border-slate-800",
    text: "text-slate-700 dark:text-slate-300",
    icon: "save",
  },
  callback: {
    bg: "bg-teal-50 dark:bg-teal-950/30",
    badge: "bg-teal-50 dark:bg-teal-950/30",
    background: "bg-teal-50 dark:bg-teal-950/30",
    border: "border-teal-200 dark:border-teal-800",
    text: "text-teal-700 dark:text-teal-300",
    icon: "phone-callback",
  },
  logger: {
    bg: "bg-lime-50 dark:bg-lime-950/30",
    badge: "bg-lime-50 dark:bg-lime-950/30",
    background: "bg-lime-50 dark:bg-lime-950/30",
    border: "border-lime-200 dark:border-lime-800",
    text: "text-lime-700 dark:text-lime-300",
    icon: "file-text",
  },
  agent: {
    bg: "bg-violet-50 dark:bg-violet-950/30",
    badge: "bg-violet-50 dark:bg-violet-950/30",
    background: "bg-violet-50 dark:bg-violet-950/30",
    border: "border-violet-200 dark:border-violet-800",
    text: "text-violet-700 dark:text-violet-300",
    icon: "user",
  },
  global: {
    bg: "bg-stone-50 dark:bg-stone-950/30",
    badge: "bg-stone-50 dark:bg-stone-950/30",
    background: "bg-stone-50 dark:bg-stone-950/30",
    border: "border-stone-200 dark:border-stone-800",
    text: "text-stone-700 dark:text-stone-300",
    icon: "globe",
  },
  unknown: {
    bg: "bg-gray-50 dark:bg-gray-950/30",
    badge: "bg-gray-50 dark:bg-gray-950/30",
    background: "bg-gray-50 dark:bg-gray-950/30",
    border: "border-gray-200 dark:border-gray-800",
    text: "text-gray-700 dark:text-gray-300",
    icon: "help-circle",
  },
};

// =============================================================================
// Flow Type Themes
// =============================================================================

export interface FlowTypeTheme {
  stroke: string;
  strokeDasharray?: string;
  label: string;
}

export const FLOW_TYPE_THEMES: Record<FlowType, FlowTypeTheme> = {
  tensor: {
    stroke: "stroke-blue-500 dark:stroke-blue-400",
    label: "Tensor",
  },
  parameters: {
    stroke: "stroke-green-500 dark:stroke-green-400",
    strokeDasharray: "5,5",
    label: "Parameters",
  },
  gradient: {
    stroke: "stroke-red-500 dark:stroke-red-400",
    strokeDasharray: "3,3",
    label: "Gradient",
  },
  control: {
    stroke: "stroke-gray-500 dark:stroke-gray-400",
    strokeDasharray: "10,5",
    label: "Control",
  },
  config: {
    stroke: "stroke-purple-500 dark:stroke-purple-400",
    strokeDasharray: "1,3",
    label: "Config",
  },
  reference: {
    stroke: "stroke-gray-500 dark:stroke-gray-400",
    label: "Reference",
  },
};

// =============================================================================
// Impl Source Themes (for value origin indicator)
// =============================================================================

export interface ImplSourceTheme {
  /** Badge background color */
  background: string;
  /** Badge text color */
  text: string;
  /** Badge border */
  border: string;
  /** Display label */
  label: string;
  /** Description for tooltip */
  description: string;
  /** Whether value is editable via YAML */
  editable: boolean;
}

export const IMPL_SOURCE_THEMES: Record<ImplSource, ImplSourceTheme> = {
  yaml: {
    background: "bg-blue-100 dark:bg-blue-900/50",
    text: "text-blue-700 dark:text-blue-300",
    border: "border-blue-300 dark:border-blue-700",
    label: "YAML",
    description: "Value from YAML config (editable)",
    editable: true,
  },
  default: {
    background: "bg-amber-100 dark:bg-amber-900/50",
    text: "text-amber-700 dark:text-amber-300",
    border: "border-amber-300 dark:border-amber-700",
    label: "Default",
    description: "Code default value (add to YAML to override)",
    editable: false,
  },
  hardcode: {
    background: "bg-red-100 dark:bg-red-900/50",
    text: "text-red-700 dark:text-red-300",
    border: "border-red-300 dark:border-red-700",
    label: "Hardcode",
    description: "Hardcoded in source code (code edit required)",
    editable: false,
  },
};

File: web_helper/backend/models/hierarchy.py
"""Hierarchical Node Graph Schema for Simulink-style visualization.

This module defines the Pydantic models for the hierarchical node graph system.
Key design principles:
1. Level 0 shows components directly (no method level)
2. Edges are extracted from both setup() and train_step()
3. No Config-based edge inference - only code-driven extraction
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field


class HierarchyLevel(str, Enum):
    """Hierarchy levels in the graph."""

    METHOD = "method"  # Level -1: Agent methods (setup, train_step, etc.)
    COMPONENT = "component"  # Level 0: setup() components
    LAYER = "layer"  # Level 1: model layers, loss sub-components
    OPERATION = "operation"  # Level 2+: internal operations


class ComponentCategory(str, Enum):
    """Component categories matching cvlabkit component types."""

    MODEL = "model"
    OPTIMIZER = "optimizer"
    SCHEDULER = "scheduler"
    LOSS = "loss"
    DATASET = "dataset"
    DATALOADER = "dataloader"
    TRANSFORM = "transform"
    METRIC = "metric"
    SAMPLER = "sampler"
    CHECKPOINT = "checkpoint"
    CALLBACK = "callback"
    LOGGER = "logger"
    AGENT = "agent"
    GLOBAL = "global"
    UNKNOWN = "unknown"


class MethodCategory(str, Enum):
    """Method categories for Agent methods."""

    SETUP = "setup"  # setup() - initialization
    TRAIN = "train"  # train_step() - training logic
    VALIDATION = "validation"  # val_step() - validation logic
    LIFECYCLE = "lifecycle"  # on_train_start, on_train_end, etc.
    UTILITY = "utility"  # Other helper methods


class FlowType(str, Enum):
    """Edge flow types based on actual code patterns."""

    TENSOR = "tensor"  # Tensor data flow (outputs, inputs)
    PARAMETERS = "parameters"  # model.parameters() -> optimizer
    GRADIENT = "gradient"  # loss.backward() gradient flow
    CONTROL = "control"  # Execution order (no data)
    CONFIG = "config"  # Config parameter injection
    REFERENCE = "reference"  # Component instance reference (self.xxx -> yyy)


class ImplSource(str, Enum):
    """Source of component implementation name.

    Used to distinguish where the impl comes from for Builder vs Execute contexts:
    - HARDCODE: impl is fixed in code, YAML is ignored
    - DEFAULT: impl is fallback when YAML doesn't specify
    - YAML: impl comes from YAML config (Builder can't determine)
    """

    HARDCODE = "hardcode"  # create.model("resnet18") - positional arg
    DEFAULT = "default"  # create.model(default="resnet18") - keyword arg
    YAML = "yaml"  # create.model() - no args, impl from YAML


class ValueSource(str, Enum):
    """Source of a property/parameter value for UI display.

    5-state system for clear visual distinction:
    - REQUIRED: No default value, must be provided (red warning)
    - CONFIG: Value loaded from YAML config (blue, editable)
    - DEFAULT: Has default value in code (gray, optional)
    - HARDCODE: Fixed in code, not configurable (gray disabled)
    - CONNECTED: Value comes from connected node port (green)
    """

    REQUIRED = "required"  # cfg.get("key") - no default, must provide
    CONFIG = "config"  # Value from YAML config file
    DEFAULT = "default"  # cfg.get("key", 0.01) - has default
    HARDCODE = "hardcode"  # Literal value in code, immutable
    CONNECTED = "connected"  # Value from connected node edge


class OriginType(str, Enum):
    LOCAL_CALL = 'local_call'
    
    """What code pattern created this node."""

    CREATE_CALL = "create_call"  # self.xxx = self.create.yyy()
    ASSIGNMENT = "assignment"  # variable assignment
    METHOD_CALL = "method_call"  # self.component(...)
    UNPACK = "unpack"  # a, b = tuple
    INPUT = "input"  # method parameter
    OUTPUT = "output"  # return value


class SourceLocation(BaseModel):
    """Source code location for traceability."""

    file: str
    line: int | None = None
    column: int = 0
    end_line: int | None = None
    end_column: int | None = None


class NodeOrigin(BaseModel):
    """Code origin information for a node."""

    type: OriginType
    create_path: list[str] | None = None  # ['model'] or ['loss', 'supervised']
    code_snippet: str
    source: SourceLocation | None = None
    impl_source: ImplSource | None = None  # YAML, DEFAULT, or HARDCODE


class DrillTarget(BaseModel):
    """Drill-down target specification."""

    type: Literal["component", "layer"]
    path: str  # Component file path or layer path


class Port(BaseModel):
    """Node port for connections."""

    name: str
    type: str = "any"  # tensor, parameters, etc.
    kind: Literal["exec", "data"] = "data"


class PropertyInfo(BaseModel):
    """Property information with value source for UI display.

    Used in SmartNodeView to show properties with clear source indicators.
    """

    name: str
    value: Any = None
    source: ValueSource = ValueSource.DEFAULT
    is_grid_search: bool = False  # True if value is array for grid search
    grid_count: int | None = None  # Number of variants if grid search
    default_value: Any = None  # Default value from code (if available)
    description: str | None = None  # Optional description/tooltip


class PropertySummary(BaseModel):
    """Summary counts of properties by source for badge display.

    Shows "⚠️ 2 required, 3 config" on node surface.
    """

    required_count: int = 0
    config_count: int = 0
    default_count: int = 0
    hardcode_count: int = 0
    connected_count: int = 0


class HierarchyPath(BaseModel):
    """Path item for breadcrumb navigation."""

    level: HierarchyLevel
    label: str
    node_id: str
    graph_id: str
    # Optional: Category for visual styling (e.g., "model", "optimizer")
    category: ComponentCategory | None = None
    # Optional: Implementation name (e.g., "resnet18")
    implementation: str | None = None


class HierarchyNode(BaseModel):
    """Node in the hierarchy graph."""

    id: str
    label: str
    level: HierarchyLevel

    # 항목 1: 객체명과 메서드명 분리
    object_name: str | None = None  # e.g., "model"
    method_name: str | None = None  # e.g., "forward"

    # Code origin (what pattern created this node?)
    origin: NodeOrigin | None = None

    # Drill-down support
    can_drill: bool = False
    drill_target: DrillTarget | None = None

    # Visual category (for coloring)
    category: ComponentCategory | None = None

    # Ports for connections
    inputs: list[Port] = Field(default_factory=list)
    outputs: list[Port] = Field(default_factory=list)

    # Source location
    source: SourceLocation | None = None

    # Properties grouped by source (for SmartNodeView)
    properties: list[PropertyInfo] = Field(default_factory=list)
    property_summary: PropertySummary | None = None

    # Used config keys (for setup phase visualization)
    used_config_keys: list[str] = Field(default_factory=list)

    # Additional metadata
    metadata: dict[str, Any] = Field(default_factory=dict)


class CodeFlowEdge(BaseModel):
    """Edge representing actual code flow."""

    id: str
    source_node: str
    source_port: str = "out"
    target_node: str
    target_port: str = "in"

    # Flow type based on code analysis
    flow_type: FlowType = FlowType.TENSOR

    edge_type: Literal["execution", "data"] | None = None

    # Variable carrying this data (from code analysis)
    variable_name: str | None = None

    # Execution order (1-based sequence index)
    sequence_index: int | None = None

    # Which method this edge was extracted from
    extracted_from: str | None = None  # "setup" or "train_step"

    # Source location
    source: SourceLocation | None = None

    # Additional metadata (is_gapped, comment, etc.)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Hierarchy(BaseModel):
    """Hierarchy context for navigation."""

    parent_graph_id: str | None = None
    parent_node_id: str | None = None
    depth: int = 0
    path: list[HierarchyPath] = Field(default_factory=list)


class HierarchicalNodeGraph(BaseModel):
    """Root schema for hierarchical node graph.

    This is the main response model for the /hierarchy/{agent} endpoint.
    """

    id: str
    label: str
    level: HierarchyLevel

    # Hierarchy context (Simulink-style navigation)
    hierarchy: Hierarchy = Field(default_factory=Hierarchy)

    # Graph content
    nodes: list[HierarchyNode] = Field(default_factory=list)
    edges: list[CodeFlowEdge] = Field(default_factory=list)

    # Agent name (for context)
    agent_name: str | None = None

    # Source file
    source_file: str | None = None

    # Code coverage info (for 1:1 code-node mapping)
    method_range: SourceLocation | None = None  # Start/end of analyzed method
    covered_lines: list[int] = Field(default_factory=list)  # Lines covered by nodes
    uncovered_lines: list[int] = Field(
        default_factory=list
    )  # Statement lines NOT covered

    # Additional metadata
    metadata: dict[str, Any] = Field(default_factory=dict)


# Response wrapper
class HierarchyResponse(BaseModel):
    """API response wrapper."""

    success: bool = True
    data: HierarchicalNodeGraph | None = None
    error: str | None = None


# =============================================================================
# Node/Edge Mutation Request/Response Models (Phase C)
# =============================================================================


class AddNodeRequest(BaseModel):
    """Request to add a new component node.

    Role (변수명) handling:
    - If role is provided, use it directly
    - If role is None, auto-generate from category (e.g., "model")
    - If duplicate, append suffix (_2, _3, etc.)
    """

    category: str  # "model", "optimizer", "loss", "dataset", "transform"
    implementation: str  # "resnet18", "adam", "cross_entropy"
    role: str | None = None  # 변수명 (없으면 category 기반 자동 생성)
    config: dict[str, Any] = Field(default_factory=dict)  # 컴포넌트 config params
    position: dict[str, float] = Field(
        default_factory=lambda: {"x": 0, "y": 0}
    )  # UI 위치


class AddNodeResponse(BaseModel):
    """Response from adding a node."""

    success: bool
    node_id: str | None = None  # 생성된 노드 ID (= role)
    role: str | None = None  # 실제 사용된 role (자동 생성된 경우 포함)
    mode: Literal["draft", "direct"] | None = None
    edit: dict[str, Any] | None = None  # GraphEdit as dict
    error: str | None = None


class UpdateNodeRequest(BaseModel):
    """Request to update an existing node."""

    implementation: str | None = None  # 새 implementation
    config: dict[str, Any] | None = None  # 변경할 config (key: value pairs)
    position: dict[str, float] | None = None  # UI 위치
    yaml_path: str | None = None  # YAML 파일 경로 (config 변경 시)
    category: str | None = None  # 컴포넌트 카테고리 (nested key 생성용)


class UpdateNodeResponse(BaseModel):
    """Response from updating a node."""

    success: bool
    node_id: str | None = None
    mode: Literal["draft", "direct"] | None = None
    edit: dict[str, Any] | None = None  # GraphEdit as dict
    error: str | None = None


class AddEdgeRequest(BaseModel):
    """Request to add a new edge (data dependency).

    Edge addition with code modification:
    - model → optimizer edge: adds self.model.parameters() to optimizer creation
    - In draft mode, only stores in memory
    - In direct mode or on commit, modifies actual code
    """

    source: str  # source node id (= role)
    target: str  # target node id (= role)
    source_port: str = "out"  # "out", "parameters", "output" 등
    target_port: str = "in"  # "in", "model", "data" 등
    flow_type: str = "reference"  # "data", "parameters", "config", "reference"


class AddEdgeResponse(BaseModel):
    """Response from adding an edge."""

    success: bool
    edge_id: str | None = None  # 생성된 엣지 ID
    mode: Literal["draft", "direct"] | None = None
    code_change: str | None = None  # 코드 수정 내용 (있는 경우)
    edit: dict[str, Any] | None = None  # GraphEdit as dict
    error: str | None = None
